mesa/space.py

```diff
@@ -548,7 +550,428 @@ def exists_empty_cells(self) -> bool:
         return len(self.empties) > 0
 
 
-class SingleGrid(_Grid):
+def is_lambda_function(function):
+    """Check if a function is a lambda function."""
+    return (
+        inspect.isfunction(function)
+        and len(inspect.signature(function).parameters) == 1
+    )
+
+
+def is_numpy_ufunc(function):
+    return isinstance(function, np.ufunc)
+
+
+def ufunc_requires_additional_input(ufunc):
+    # NumPy ufuncs have a 'nargs' attribute indicating the number of input arguments
+    # For binary ufuncs (like np.add), nargs is 2
+    return ufunc.nargs > 1
+
+
+class PropertyLayer:
+    """
+    A class representing a layer of properties in a two-dimensional grid. Each cell in the grid
+    can store a value of a specified data type.
+
+    Attributes:
+        name (str): The name of the property layer.
+        width (int): The width of the grid (number of columns).
+        height (int): The height of the grid (number of rows).
+        data (numpy.ndarray): A NumPy array representing the grid data.
+
+    Methods:
+        set_cell(position, value): Sets the value of a single cell.
+        set_cells(value, condition): Sets the values of multiple cells, optionally based on a condition.
+        modify_cell(position, operation, value): Modifies the value of a single cell using an operation.
+        modify_cells(operation, value, condition_function): Modifies the values of multiple cells using an operation.
+        select_cells(condition, return_list): Selects cells that meet a specified condition.
+        aggregate_property(operation): Performs an aggregate operation over all cells.
+    """
+
+    agentset_experimental_warning_given = False
+
+    def __init__(
+        self, name: str, width: int, height: int, default_value, dtype=np.float32
+    ):
+        """
+        Initializes a new PropertyLayer instance.
+
+        Args:
+            name (str): The name of the property layer.
+            width (int): The width of the grid (number of columns). Must be a positive integer.
+            height (int): The height of the grid (number of rows). Must be a positive integer.
+            default_value: The default value to initialize each cell in the grid. Should ideally
+                           be of the same type as specified by the dtype parameter.
+            dtype (data-type, optional): The desired data-type for the grid's elements. Default is np.float32.
```

Comment:
`np.float64` as the default would make more sense, as it is the default for Python's `float`.
```
In [4]: %timeit data = np.full((1000, 1000), 10, dtype=np.float32)
339 µs ± 998 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

In [5]: %timeit data = np.full((1000, 1000), 10, dtype=np.float64)
813 µs ± 6.21 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```
It's ~2x slower, but the vectorized version would still be much faster in https://github.com/projectmesa/mesa/pull/1898#issuecomment-1849000346.

Reply:
I think for much more users the extra performance wil, be notable, and I don’t see any current models where float64 is needed.

We could also make `dtype` an required argument (without default).

Reply:
> I think for much more users the extra performance wil, be notable, and I don’t see any current models where float64 is needed.

I'd be surprised if the default is `float32` in this case. Because I think the precision used should be consistent regarding with the floating point error accummulation, either `float32` everywhere, or `float64` everywhere. Users who want the speedup at the tradeoff of using `float32`, or using a mix of both, can manually choose to do so. We can document this in the [how-to guide](https://github.com/projectmesa/mesa/blob/main/docs/howto.rst).

> We could also make dtype an required argument (without default).

I'd say it should be optional, much like how it is for the `np.ndarray` constructors in NumPy.

Reply:
@Corvince @jackiekazil @tpike3 @wang-boyu @quaquel I would like additional opinions on this.

I’m leaning towards making `dtype` a required argument so that users make a conscious choice. However, that requires familiarity with NumPy’s [data types](https://numpy.org/doc/stable/user/basics.types.html) for a proper choice.

Also I’m afraid users will, just pass `float`, which equates `np.float64` to making their grid twice as slow, while `np.float32` suffices in almost every situation. I can’t think of a single model or ABM use case in which 64-bit floating point precision is required, so I don’t think it would be a good default.

Reply:
I agree with @rht. Just follow numpy and use float 64 as default. Having `dtype` as an keyword argument would be convenient if users have good reasons to deviate they can easily do so.

Reply:
https://github.com/projectmesa/mesa/pull/1898/commits/c71c9f54722c0a24bbbfb37248b7209c283efd44